crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -139,6 +253,15 @@ impl<'db> TypeInferenceBuilder<'db> {
     }
 
     fn infer_function_definition_statement(&mut self, function: &ast::StmtFunctionDef) {
+        let definition = self.index.definition(function);
+        self.infer_function_definition(function, definition);
+    }
+
+    fn infer_function_definition(
```

Comment:
I don't fully understand how the different scopes are merged, or how we avoid performing the same type inference multiple times. 

Let's say we call type inferences on module. The builder runs over each statement, calls `infer_statement` and we ultimately arrive here, performing the inference. 

Now, some other code calls `infer_definition_ty(function_stmt)`. I would expect that this re-runs the type-inference for the entire function because the module-level type inference never called the salsa query.

Reply:
Yes, you're totally right. The call above to `self.infer_function_definition` needs to go through the query instead, and then we need to merge the types from that query result. I just forgot to do this, and then no tests caught it. I'll add a test that would catch it, and fix it.